terminology/src/main/java/au/csiro/pathling/terminology/caching/CachingTerminologyService.java

```diff
@@ -0,0 +1,289 @@
+        cacheManager, LOOKUP_CACHE_NAME);
+  }
+
+  @Override
+  public boolean validateCode(@Nonnull final String valueSetUrl, @Nonnull final Coding coding) {
+    final ValidateCodeParameters parameters = new ValidateCodeParameters(valueSetUrl,
+        ImmutableCoding.of(coding));
+    final ValidateCodeExecutor executor = new ValidateCodeExecutor(terminologyClient, parameters);
+    return getFromCache(validateCodeCache, parameters, executor);
+  }
+
+  @Nonnull
+  @Override
+  public List<Translation> translate(@Nonnull final Coding coding,
+      @Nonnull final String conceptMapUrl,
+      final boolean reverse, @Nullable final String target) {
+    final TranslateParameters parameters = new TranslateParameters(ImmutableCoding.of(coding),
+        conceptMapUrl, reverse, target);
+    final TranslateExecutor executor = new TranslateExecutor(terminologyClient, parameters);
+    return getFromCache(translateCache, parameters, executor);
+  }
+
+  @Nonnull
+  @Override
+  public ConceptSubsumptionOutcome subsumes(@Nonnull final Coding codingA,
+      @Nonnull final Coding codingB) {
+    final SubsumesParameters parameters = new SubsumesParameters(
+        ImmutableCoding.of(codingA), ImmutableCoding.of(codingB));
+    final SubsumesExecutor executor = new SubsumesExecutor(terminologyClient, parameters);
+    return getFromCache(subsumesCache, parameters, executor);
+  }
+
+  @Nonnull
+  @Override
+  public List<PropertyOrDesignation> lookup(@Nonnull final Coding coding,
+      @Nullable final String property) {
+    final LookupParameters parameters = new LookupParameters(ImmutableCoding.of(coding), property);
+    final LookupExecutor executor = new LookupExecutor(terminologyClient, parameters);
+    return getFromCache(lookupCache, parameters, executor);
+  }
+
+  /**
+   * Gets the result of an operation from the cache, or fetches a new result if the cache is empty
+   * or expired.
+   *
+   * @param cache The cache being used for this operation
+   * @param parameters The parameters for the operation
+   * @param operation A {@link TerminologyOperation} that provides the behavior specific to the type
+   * of operation
+   * @param <ParametersType> The type of the parameters that are input to the operation
+   * @param <ResponseType> The type of the response returned by the terminology client
+   * @param <ResultType> The type of the final result that is extracted from the response
+   * @return The operation result
+   */
+  private <ParametersType extends TerminologyParameters, ResponseType, ResultType extends Serializable> ResultType getFromCache(
+      @Nonnull final Cache<Integer, TerminologyResult<ResultType>> cache,
+      @Nonnull final ParametersType parameters,
+      @Nonnull final TerminologyOperation<ResponseType, ResultType> operation
+  ) {
+    final int key = parameters.hashCode();
+    final TerminologyResult<ResultType> cached = cache.get(key);
+
```

Comment:
> I think it would be better to use `parameters` as the key rather than their hashCodes(), as using hashKey() created potential for collisions.

I don't really understand this point, would you mind elaborating on this a little? Wouldn't the `hashCode()` be used anyway, if `TerminologyParameters` was the key? What about persistence, would `TerminologyParameters` get serialized, or would just its hash code get persisted?

Reply:
> And it either case, the `parameters` as such as suboptimal keys, as they include full codings, while only system, code and version are ever used in terminology calls. (so 1000 codings that differ only in id or display result in 1000 cache misses).

I think this could be fixed by overriding the `hashCode()` function for `ImmutableCoding`, to only include system, code and version?

Reply:
I've made this change.

I've also been thinking about some sort of cache stats collector, to check and analyse the hit rate.